src/sql/workbench/parts/objectExplorer/common/objectExplorerViewTreeShim.ts

```diff
@@ -44,45 +45,56 @@ export class OEShimService extends Disposable implements IOEShimService {
 		if (this.cm.providerRegistered(providerId)) {
-			await new Promise(async (resolve, reject) => {
-				await this.cm.connect(connProfile, undefined,
-					{ showConnectionDialogOnError: true, showFirewallRuleOnError: true, saveTheConnection: false, showDashboard: false, params: undefined },
-					{
-						onConnectSuccess: async (e, profile) => {
-							let existingConnection = this.cm.findExistingConnection(profile);
-							connProfile = new ConnectionProfile(this.capabilities, await this.cm.addSavedPassword(existingConnection));
-							resolve();
-						},
-						onConnectCanceled: () => {
-							reject('User canceled');
-						},
-						onConnectReject: undefined,
-						onConnectStart: undefined,
-						onDisconnect: undefined
-					});
-			});
+			connProfile = await this.connectOrPrompt(connProfile);
+		} else {
+			// Throw and expect upstream handler to notify about the error
+			// TODO: In the future should use extension recommendations to prompt for correct extension
+			throw new Error(localize('noProviderFound', "Cannot expand as the required connection provider '{0}' was not found", providerId));
 		}
 		let sessionResp = await this.oe.createNewSession(providerId, connProfile);
-		let disp = this.oe.onUpdateObjectExplorerNodes(e => {
-			if (e.connection.id === connProfile.id) {
-				if (e.errorMessage) {
-					deferred.reject();
-					return;
-				}
-				let rootNode = this.oe.getSession(sessionResp.sessionId).rootNode;
-				// this is how we know it was shimmed
-				if (rootNode.nodePath) {
-					this.nodeHandleMap.set(generateNodeMapKey(viewId, node), rootNode.nodePath);
+		let sessionId = sessionResp.sessionId;
+		await new Promise((resolve, reject) => {
+			let listener = this.oe.onUpdateObjectExplorerNodes(e => {
+				if (e.connection.id === connProfile.id) {
+					if (e.errorMessage) {
+						listener.dispose();
+						reject(new Error(e.errorMessage));
+						return;
+					}
+					let rootNode = this.oe.getSession(sessionResp.sessionId).rootNode;
+					// this is how we know it was shimmed
+					if (rootNode.nodePath) {
+						this.nodeHandleMap.set(generateNodeMapKey(viewId, node), rootNode.nodePath);
+					}
 				}
-			}
-			disp.dispose();
-			deferred.resolve(sessionResp.sessionId);
+				listener.dispose();
+				resolve(sessionResp.sessionId);
+			});
+		});
+		return sessionId;
+	}
+
+	private async connectOrPrompt(connProfile: ConnectionProfile): Promise<ConnectionProfile> {
+		connProfile = await new Promise(async (resolve, reject) => {
+			await this.cm.connect(connProfile, undefined, { showConnectionDialogOnError: true, showFirewallRuleOnError: true, saveTheConnection: false, showDashboard: false, params: undefined }, {
+				onConnectSuccess: async (e, profile) => {
+					let existingConnection = this.cm.findExistingConnection(profile);
+					connProfile = new ConnectionProfile(this.capabilities, existingConnection);
+					connProfile = <ConnectionProfile>await this.cm.addSavedPassword(connProfile);
```

Comment:
Is this the only actual change? (Looks like a bunch of code moving around).

Reply:
How is this actually fixing the issue? We were calling addSavedPassword before, just directly as a param to the ConnectionProfile constructor, so I'm not sure how this is actually changing anything

---
In reply to: [289190793](https://github.com/microsoft/azuredatastudio/pull/5768#discussion_r289190793) [](ancestors = 289190793)

Reply:
You called addSavedPassword on the existing connection, then copied it to connProfile which caused it to be lost. Instead we copy to connProfile and addSavedPassword on that which adds back in the data needed.

Reply:
And yes - this is the primary fix. I think that moving to a consistent model of just using promises instead of having multiple different methods (Deferred) which aren't needed is a helpful improvement. It's strange to have a deferred in an async method

Reply:
Ah, so yeah ConnectionProfile isn't copying over the password (or honestly most stuff like server, db, etc). Not what I'd expect from how it was called but yeah your fix makes sense.